src/lib.rs

```diff
@@ -78,3 +78,62 @@ pub fn from_value<T: serde::de::DeserializeOwned>(value: JsValue) -> Result<T> {
 pub fn to_value<T: serde::ser::Serialize + ?Sized>(value: &T) -> Result<JsValue> {
     value.serialize(&Serializer::new())
 }
+
+/// A wrapper around a [`JsValue`] (or anything that can be cast into one)
+/// that makes it pass through serialization and deserialization unchanged.
+///
+/// # Example
+/// ```rust
+/// use serde_wasm_bindgen::{PreservedValue, to_value};
+/// #[derive(serde::Serialize)]
+/// struct MyStruct {
+///     int_field: i32,
+///     js_field: PreservedValue<JsValue>,
+/// }
+/// let big_array = js_sys::Int8Array::new_with_length(1000);
+/// let s = MyStruct {
+///     int_field: 5,
+///     js_field: PreservedValue(big_array.into()),
+/// };
+///
+/// // Will return a JsValue representing an object with two fields (`int_field` and `js_field`).
+/// // `js_field` will be an `Int8Array` pointing to the same underlying JavaScript object as `big_array`.
+/// to_value(&s);
+/// ```
+#[derive(Clone, Debug, PartialEq, serde::Serialize, serde::Deserialize)]
+#[serde(bound = "T: JsCast")]
+pub struct PreservedValue<T: JsCast>(#[serde(with = "preserved_value")] pub T);
+
+// Some arbitrary string that no one will collide with unless they try.
+pub(crate) const PRESERVED_VALUE_MAGIC: &str = "1fc430ca-5b7f-4295-92de-33cf2b145d38";
+
+mod preserved_value {
+    use serde::{de::Error, Deserialize, Serialize};
+    use wasm_bindgen::{
+        convert::{FromWasmAbi, IntoWasmAbi},
+        JsCast, JsValue,
+    };
+
+    #[derive(Serialize, Deserialize)]
+    #[serde(rename = "1fc430ca-5b7f-4295-92de-33cf2b145d38")]
+    struct PreservedValueWrapper(u32);
+
+    pub fn serialize<S: serde::Serializer, T: JsCast>(val: &T, ser: S) -> Result<S::Ok, S::Error> {
+        // Since we don't own `val` we need to clone it. Otherwise serializing a JsValue
+        // will produce a second JsValue pointing to the same index in the wasm-bindgen heap,
+        // and whichever one is dropped first will leave the other one dangling.
+        PreservedValueWrapper(val.as_ref().clone().into_abi()).serialize(ser)
+    }
+
+    pub fn deserialize<'de, D: serde::Deserializer<'de>, T: JsCast>(de: D) -> Result<T, D::Error> {
+        let wrap = PreservedValueWrapper::deserialize(de)?;
+        // When used with our deserializer this unsafe is correct, because the
+        // deserializer just converted a JsValue into_abi.
+        // With other deserializers, this may be incorrect but it shouldn't be UB
+        // because JsValues are represented using indices into a JS-side (i.e.
+        // bounds-checked) array.
+        let val: JsValue = unsafe { FromWasmAbi::from_abi(wrap.0) };
+        val.dyn_into()
+            .map_err(|e| D::Error::custom(format!("incompatible JS value {e:?}")))
```

Comment:
```suggestion
            .map_err(|e| D::Error::custom(format_args!("incompatible JS value {e:?}")))
```

This avoids one intermediate allocation. Since `custom` itself accepts any `Display` and converts it to an owned String, passing an owned String to it is a little bit suboptimal.

Reply:
Perhaps also worth using [`std::any::type_name`](https://doc.rust-lang.org/std/any/fn.type_name.html) to tell user which type was, in fact, expected.